### PR TITLE
[DB Change] Support team disqualification

### PR DESCRIPTION
**Pull request by lambertwang**

### CODE REVIEW

File: ServerCore/Pages/Submissions/Index.cshtml [hunk 1]
```diff
@@ -11,11 +11,14 @@
 
 @{
     var mailtoUrl = "mailto:";
-    
+
     // to: Add the proper support agent as recipient
     mailtoUrl += Model.PuzzleState?.Puzzle.SupportEmailAlias ?? (Model.Event?.ContactEmail ?? "puzzhunt@microsoft.com");
     // cc: Cc the team's email address because what if this email account is a random personal one?
-    mailtoUrl += "?cc=" + Uri.EscapeDataString(Model.Team.PrimaryContactEmail);
+    if (Model.Team.PrimaryContactEmail != null)
```
Comment (tabascq) [inline ServerCore/Pages/Submissions/Index.cshtml]: pretty sure this mail is only null for demo teams but this is harmless so no big deal.

File: ServerCore/Pages/Teams/AllTeams.cshtml [hunk 2]
```diff
@@ -37,6 +44,20 @@ Registered Teams: <b>@Model.Teams.Count/@Model.Event.MaxNumberOfTeams</b>, Regis
                 <td>
                     @Html.DisplayFor(modelItem => @Model.PlayerCountByTeamID[item.ID])
                 </td>
+                @if (Model.EventRole == ModelBases.EventRole.admin ||
+                     Model.EventRole == ModelBases.EventRole.author)
+                {
+                    if (item.IsDisqualified)
+                    {
+                        <th>
```
Comment (morganbr) [inline ServerCore/Pages/Teams/AllTeams.cshtml]: This should be ```<td>```

File: ServerCore/Pages/Teams/AllTeams.cshtml [hunk 3]
```diff
@@ -37,6 +44,20 @@ Registered Teams: <b>@Model.Teams.Count/@Model.Event.MaxNumberOfTeams</b>, Regis
...
+                            Disqualified
+                        </th>
+                    }
+                    else
+                    {
+                        <th></th>
```
Comment (morganbr) [inline ServerCore/Pages/Teams/AllTeams.cshtml]: ```<td></td>```

File: ServerCore/Pages/Teams/Details.cshtml [hunk 4]
```diff
@@ -10,6 +10,30 @@
...
+
+    // to: Add the proper support agent as recipient
+    mailtoUrl += Model.Event?.ContactEmail ?? "puzzhunt@microsoft.com";
+    // cc: Cc the team's email address because what if this email account is a random personal one?
+    if (Model.Team.PrimaryContactEmail != null)
+    {
+        mailtoUrl += "?cc=" + Uri.EscapeDataString(Model.Team.PrimaryContactEmail);
+    }
+
+    // subject: Make this be about this puzzle
```
Comment (morganbr) [inline ServerCore/Pages/Teams/Details.cshtml]: This comment doesn't look accurate

File: ServerCore/Pages/Teams/Details.cshtml [hunk 5]
```diff
@@ -10,6 +10,30 @@
...
+    mailtoUrl += "&subject=" + Uri.EscapeDataString("[" + Model.Event.Name + "]");
+    // subject: Make this be from this team
+    mailtoUrl += Uri.EscapeDataString(" [" + Model.Team.Name + "]");
+
+    // request out of email mode
```
Comment (morganbr) [inline ServerCore/Pages/Teams/Details.cshtml]: Ditto

File: ServerCore/Pages/Teams/Details.cshtml [hunk 6]
```diff
@@ -10,6 +10,30 @@
 }
 @using Microsoft.AspNetCore.Http.Extensions @* for GetEncodedUrl *@
 
+@{
+    var mailtoUrl = "mailto:";
```
Comment (morganbr) [inline ServerCore/Pages/Teams/Details.cshtml]: Could this logic be under the disqualification check so we don't have to build the email for the vast majority of cases where the team isn't disqualified?

Comment (tabascq) [inline ServerCore/Pages/Teams/Details.cshtml]: Possible to make this a StringBuilder and Append() all these things? It's somewhat more efficient to do it that way.

File: ServerCore/Pages/Teams/Details.cshtml [hunk 7]
```diff
@@ -10,6 +10,30 @@
...
+    mailtoUrl += "&subject=" + Uri.EscapeDataString("[" + Model.Event.Name + "]");
+    // subject: Make this be from this team
+    mailtoUrl += Uri.EscapeDataString(" [" + Model.Team.Name + "]");
+
+    // request out of email mode
+    // subject: Add the Email Mode signifier
+    mailtoUrl += Uri.EscapeDataString(" [⚠ Team Disqualified ⚠]");
+    // body: Invite solver to give details
+    mailtoUrl += "&body=" + Uri.EscapeDataString(
+        "Use this email to fully explain your thought process so we know why you would like your team disqualification to be undone! 😇" + Environment.NewLine + Environment.NewLine);
```
Comment (tabascq) [inline ServerCore/Pages/Teams/Details.cshtml]: "Use this mail to explain anything you think we need to know." As written, it sounds like we'll happily undisqualify you for any reason...

File: ServerCore/Pages/Teams/Details.cshtml [hunk 8]
```diff
@@ -27,15 +51,49 @@
...
 <h2>@Html.DisplayFor(model => model.Team.Name) Team</h2>
 
+@if (Model.Team.IsDisqualified)
+{
+    <div class="alert alert-danger" role="alert">
+        Your team has been disqualified from the event. You are welcome to
+        continue solving puzzles for fun, but your team will no longer appear
+        in the final standings and is not eligible for any awards or prizes
+        that the event may offer. If you have any questions, please contact 
+        <a href="@mailtoUrl">.
```
Comment (morganbr) [inline ServerCore/Pages/Teams/Details.cshtml]: It looks like this tag isn't wrapping anything and isn't closed.

File: ServerCore/Pages/Teams/Details.cshtml [hunk 9]
```diff
@@ -27,15 +51,49 @@
...
+    </div>
+}
+
 @if (Model.EventRole != ModelBases.EventRole.author)
 {
     <div>
-        <a asp-page="./Edit" asp-route-teamId="@Model.Team.ID">Edit team details</a> |
+        <a asp-page="./Edit" asp-route-teamId="@Model.Team.ID">Edit team details</a>
+        |
```
Comment (tabascq) [inline ServerCore/Pages/Teams/Details.cshtml]: What are these extra lines for?

File: ServerCore/Pages/Teams/Disqualify.cshtml [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+    <dl>
+        <!-- class="dl-horizontal"> -->
+        <dt>
+            Team name
+        </dt>
+        <dd>
+            @Html.DisplayFor(model => model.Team.Name)
+        </dd>
+        <!-- TODO: Conditionally show/hide based on event property (not needed for PuzzleHunt)
+    <dt>
```
Comment (morganbr) [inline ServerCore/Pages/Teams/Disqualify.cshtml]: I'm not sure we need all of the details about the team here. Name and perhaps contact info seems like enough since we can use the details page for the rest

File: ServerCore/Pages/Teams/Details.cshtml [hunk 11]
```diff
@@ -27,15 +51,49 @@
...
         }
         <a asp-page="./Delete" asp-route-teamId="@Model.Team.ID">Delete team</a>
+        |
+        @if (Model.EventRole == ModelBases.EventRole.admin)
+        {
+            @if (@Model.Team.IsDisqualified)
+            {
+                <a asp-page="./Disqualify" 
+                   asp-route-teamId="@Model.Team.ID">
+                    Requalify Team
```
Comment (tabascq) [inline ServerCore/Pages/Teams/Details.cshtml]: @(Model.Team.IsDisqualified ? "Requalify Team" : "Disqualify Team") - then you don't need to repeat the entire <a> tag.

File: ServerCore/Pages/Teams/Disqualify.cshtml [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+    <hr />
+    <dl>
+        <!-- class="dl-horizontal"> -->
+        <dt>
+            Team name
+        </dt>
+        <dd>
+            @Html.DisplayFor(model => model.Team.Name)
+        </dd>
+        <!-- TODO: Conditionally show/hide based on event property (not needed for PuzzleHunt)
```
Comment (tabascq) [inline ServerCore/Pages/Teams/Disqualify.cshtml]: remove this and many properties below

File: ServerCore/Pages/Teams/Disqualify.cshtml.cs [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+                .Select(tm => tm.Member.Email)
+                .ToListAsync();
+
+            if (MailSubject?.Length > 0)
+            {
+                MailHelper.Singleton.SendPlaintextBcc(addresses, MailSubject, MailBody);
+            }
+            else
+            {
+                var mailtoUrl = "mailto:";
```
Comment (morganbr) [inline ServerCore/Pages/Teams/Disqualify.cshtml.cs]: It doesn't look like mailToUrl gets used for anything

File: ServerCore/Pages/Teams/Disqualify.cshtml.cs [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+        {
+            Team = await _context.Teams.FindAsync(teamId);
+
+            if (Team != null)
+            {
+                await TeamHelper.SetTeamQualificationAsync(_context, Team, true);
+            }
+
+            IEnumerable<string> addresses = Enumerable.Empty<string>();
+            addresses = await _context.TeamMembers
```
Comment (tabascq) [inline ServerCore/Pages/Teams/Disqualify.cshtml.cs]: This is the entire event!!! Filter on team.

File: ServerCore/Pages/Teams/Disqualify.cshtml.cs [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+                    "[" + Team.Name + "] Disqualified from event";
+
+                MailHelper.Singleton.SendPlaintextBcc(
+                    addresses,
+                    DefaultMailSubject,
+                    DefaultMailBody);
+
+            }
+
+            if (EventRole == EventRole.admin)
```
Comment (tabascq) [inline ServerCore/Pages/Teams/Disqualify.cshtml.cs]: ??? Only admins should be able to get here.

File: ServerCore/Pages/Teams/Status.cshtml.cs [hunk 16]
```diff
@@ -83,5 +84,12 @@ public async Task<IActionResult> OnGetEmailModeAsync(int teamId, int? puzzleId,
             // redirect without the solve info to keep the URL clean
             return RedirectToPage(new { teamId, sort });
         }
+
+        public async Task<IActionResult> OnGetDisqualifyAsync(Team team, bool value, string sort)
```
Comment (tabascq) [inline ServerCore/Pages/Teams/Status.cshtml.cs]: When is this used? Seems dangerous?

File: ServerCore/Pages/Submissions/Index.cshtml (hunk 1 above)
Comment (lambertwang) [reply ServerCore/Pages/Submissions/Index.cshtml]: Okay, I just added it since I was getting errors while testing this with the default sample event.

File: ServerCore/Pages/Submissions/Index.cshtml [hunk 17]
```diff
@@ -11,11 +11,14 @@
 
 @{
     var mailtoUrl = "mailto:";
```
Comment (tabascq) [inline ServerCore/Pages/Submissions/Index.cshtml]: could StringBuilder this part too, for perf

File: ServerCore/Helpers/TeamHelper.cs [hunk 18]
```diff
@@ -148,5 +148,16 @@ public static bool IsMicrosoftNonIntern(string email)
...
             return email.EndsWith("@microsoft.com") && !email.StartsWith("t-");
         }
+
+        public static async Task SetTeamQualificationAsync(
+            PuzzleServerContext context,
+            Team team,
+            bool value)
+
+        {
+
```
Comment (morganbr) [inline ServerCore/Helpers/TeamHelper.cs]: Extra empty line

File: ServerCore/Pages/Teams/AllTeams.cshtml [hunk 19]
```diff
@@ -37,6 +44,20 @@ Registered Teams: <b>@Model.Teams.Count/@Model.Event.MaxNumberOfTeams</b>, Regis
                 <td>
                     @Html.DisplayFor(modelItem => @Model.PlayerCountByTeamID[item.ID])
                 </td>
+                @if (Model.EventRole == ModelBases.EventRole.admin ||
+                     Model.EventRole == ModelBases.EventRole.author)
+                {
+                    if (item.IsDisqualified)
+                    {
+                        <td>
```
Comment (jenetlan) [inline ServerCore/Pages/Teams/AllTeams.cshtml]: Seems like you could put the entire if statement inside the <td> so that it's obvious that this section is filling out the single cell?

File: ServerCore/Pages/Submissions/Index.cshtml [hunk 20]
```diff
@@ -9,36 +9,41 @@
...
+
 @{
-    var mailtoUrl = "mailto:";
+    var mailtoUrl = new StringBuilder("mailto:");
 
     // to: Add the proper support agent as recipient
-    mailtoUrl += Model.PuzzleState?.Puzzle.SupportEmailAlias ?? (Model.Event?.ContactEmail ?? "puzzhunt@microsoft.com");
-
+    mailtoUrl.Append(Model.PuzzleState?.Puzzle.SupportEmailAlias ?? (Model.Event?.ContactEmail ?? "puzzhunt@microsoft.com"));
+    // cc: Cc the team's email address because what if this email account is a random personal one?
```
Comment (jenetlan) [inline ServerCore/Pages/Submissions/Index.cshtml]: Maybe rephrase to "CC the team's email address in case the account sending the e-mail is a personal one that's not associated with the event"?

File: ServerCore/Pages/Teams/Details.cshtml [hunk 21]
```diff
@@ -9,6 +9,38 @@
...
+@using System.Text @* for StringBuilder *@
+
+@{
+    var mailtoUrl = new StringBuilder("mailto:");
+
+    if (Model.Team.IsDisqualified)
+    {
+        // to: Add the proper support agent as recipient
+        mailtoUrl.Append(Model.Event?.ContactEmail ?? "puzzhunt@microsoft.com");
+        // cc: Cc the team's email address because what if this email account is a random personal one?
```
Comment (jenetlan) [inline ServerCore/Pages/Teams/Details.cshtml]: Same comment as above wrt comment rephrasing

**Comment by jenetlan** [comment]
@lambertwang FYI we decided tonight that this was a bit risky to take right before the event but we're hoping to have it on deck so that if we need it for PuzzleDay we can pull it in between the end of the event and the start of closing. Do you think you can address the merge conflicts so that it's ready to go if we need it?

**Comment by jenetlan** [comment]
(also fyi updated the title to reflect that it involves a DB change - that way we don't accidentally deploy without updating the database once we take it)

**Comment by jenetlan** [comment]
Thank you! 😄
